src/daft-local-execution/src/pipeline.rs

```diff
@@ -319,18 +319,32 @@ pub fn physical_plan_to_pipeline(
             null_equals_null,
             join_type,
             schema,
+            ..
         }) => {
             let left_schema = left.schema();
             let right_schema = right.schema();
 
-            // Determine the build and probe sides based on the join type
-            // Currently it is a naive determination, in the future we should leverage the cardinality of the tables
-            // to determine the build and probe sides
+            // TODO(desmond): Stats should always be materialized by this point. We should fix this on the Daft Connect side.
+            // If stats are materialized, use them to determine the build and probe sides.
+            let left_stats_state = left.get_stats_state();
+            let right_stats_state = right.get_stats_state();
+            let left_smaller_than_right = match (left_stats_state, right_stats_state) {
+                (StatsState::Materialized(left_stats), StatsState::Materialized(right_stats)) => {
+                    left_stats.approx_stats.upper_bound_bytes
+                        <= right_stats.approx_stats.upper_bound_bytes
+                }
+                _ => true,
+            };
+
             let build_on_left = match join_type {
-                JoinType::Inner => true,
-                JoinType::Right => true,
-                JoinType::Outer => true,
+                JoinType::Inner => left_smaller_than_right,
+                JoinType::Outer => left_smaller_than_right,
+                // TODO(desmond): We might potentially want to flip the probe table side for
```

Comment:
It's "sort of" true. "Sort of" because we actually need a new algorithm for this case.

We can consider a right outer join, where the left side is a few orders of magnitude larger than the right side (e.g. left has 1B rows, right side has 10 rows. Extreme but just for illustration).

### Normal hash join
We could either do the default algorithm, build a giant probe table on the left, then stream the right's 10 records against the probe table.

### Two pass hash join

Alternatively, we can build a probe table of the right, but additionally track which rows were matched against at least one row on the left.

After streaming all rows on the left, we then check for rows on the right that were unmatched and emit them.

Reply:
Clarified the comment